### PR TITLE
EOPatch input

### PR DESCRIPTION
**Pull request by lojzezust**
## EOPatch reading

Implemented method `eopatch_dataset` for reading selected features from EOPatches into a tensorflow Dataset.

## Dataset transformations

Also added common dataset transformations: subpatch extraction, data augmentation, caching.

`extract_subpatches` creates a transformation that extracts multiple subpatches of given size from each patch in the dataset. Subpatches can be sampled randomly or from a grid.

`augment_data` creates a transformation that randomly augments each example with specified operations. Operations include rotation, flipping, brightness and contrast adjustment.

`cache_dataset` caches the content of the dataset to disk. This enables us to cache extracted patches. In this way EOPatch reading and subpatch extraction is only done during the first epoch. For the rest of training, cached data is used.

## EOPatch input method

`EOPatchInputExample` contains an example implementation of a input method that can be used in the configuration. The method shows a basic configurable dataset pipeline implementation.

## Models update

FCN, TFCN, RFCN models have been updated to recieve the labels of the same size as the input. Cropping is done internaly. Basic image summaries (input image, labels, predictions) have been added to these models as well.

Examples have been updated to fit the new model definition.

## EOPatch training example

A simple example that works on Slovenian LULC patches has been added. It showcases eopatch loading, subpatch extraction, caching and data augmentation. In order for the cache to not take up too much space, only 5 128x128 subpatches are extracted from each patch. The example can be run using the following command:

```bash
python -m eoflow.execute configs/tfcn_example_real.json 
```

Logs and summaries are saved into `temp/tfcn_svn_lulc` dir. The summaries can be visualized using the following command:

```bash
tensorboard --logdir temp/tfcn_svn_lulc
```

### CODE REVIEW

**Comment by lojzezust** [comment]
I moved the bulk of the training code inside the `BaseModel` class. This enables better programatic use in notebooks etc. `TrainTask` is now just a configurable wrapper around the `train` function of the model.

So a workflow in code would look like this:

You initialize a model object providing a dict of parameters. Then a `dataset_fn`, a function that builds the tf.dataset must be provided. For this also a `get_dataset` method of our input classes can be used, but you can also create a custom one by hand. Then `model.train(dataset_fn, ...)` can be called for training.

If this seems ok, I will do a similar thing for prediction. On the same model you will call the `predict` method with another dataset_fn as an input.

File: eoflow/base/base_model.py [hunk 1]
```diff
@@ -111,3 +84,94 @@ def build_model(self, features, labels, mode):
...
+                            # Compute and record summaries every summary_steps
+                            if training_step % summary_steps == 0:
+                                _, loss, step, summaries = sess.run([train_op, loss_op, step_tensor, summaries_op])
+
+                                summary_writer.add_summary(summaries, global_step=step)
+                            else:
+                                _, loss, step = sess.run([train_op, loss_op, step_tensor])
+
+                            # Show progress
+                            if training_step % progress_steps == 0:
```
Comment (devisperessutti) [inline eoflow/base/base_model.py]: it would be nice to use a progress bar such as the one provided by `tqdm`, and adding the loss value to the progress bar using `pbar.set_description`

File: eoflow/input/eopatch.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    def get_dataset(self):
+        cfg = self.config
+
+        # Create a tf.data.Dataset from EOPatches
+        features_data = [
+            (cfg.input_feature_type, cfg.input_feature_name, 'features', np.float32, self._parse_shape(cfg.input_feature_shape)),
+            (cfg.labels_feature_type, cfg.labels_feature_name, 'labels', np.int64, self._parse_shape(cfg.labels_feature_shape))
+        ]
+        dataset = eopatch_dataset(self.config.data_dir, features_data, fill_na=-2)
```
Comment (devisperessutti) [inline eoflow/input/eopatch.py]: This could be made a general class to interface EOPatches, by exposing some other parameter such as `fill_na`, `random_sampling` and the params required by the grid splitting
